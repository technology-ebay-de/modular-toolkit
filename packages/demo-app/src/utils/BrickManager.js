import { registerSelectorsForUseWithGlobalState } from '@modular-toolkit/selectors';

export default class {
    constructor({ store, reducer = s => s, sagaMiddleware }) {
        this.store = store;
        this.initialReducer = reducer;
        this.storePaths = {};
        this.reducers = {};
        this.sagaMiddleware = sagaMiddleware;
    }

    installBricks(bricks) {
        for (const [storePath, module] of Object.entries(bricks)) {
            this.installBrick(storePath, module);
        }
    }

    installBrick(storePath, { reducer, saga, selectors }) {
        registerSelectorsForUseWithGlobalState(storePath, selectors);
        this.addReducer(storePath, reducer);
        this.sagaMiddleware.run(saga);
    }

    addStorePath(storePathName) {
        this.storePaths[storePathName] = (state = {}, action) => {
            const reducer = this.retrieveReducer(storePathName);
            return reducer ? reducer(state, action) : state;
        };
    }

    addReducer(storePathName, reducer) {
        this.addStorePath(storePathName);
        this.storeReducer(storePathName, reducer);

        this.store.replaceReducer((state = {}, action) => {
            let nextState = { ...state, ...this.initialReducer(state, action) };
            let hasChanges = state !== nextState;
            nextState = this.addValueByDottedPath(nextState, storePathName);
            let curr = nextState;
            this.forEachPathSegment(storePathName, (segment, isLast, currPath) => {
                const objectOrReducer = this.retrieveReducer(currPath);
                if (typeof objectOrReducer === 'function') {
                    const nextSegmentState = objectOrReducer(curr[segment], action);
                    if (nextSegmentState !== curr[segment]) {
                        curr[segment] = nextSegmentState;
                        hasChanges = true;
                    }
                }
                curr = curr[segment];
            });
            return hasChanges ? nextState : state;
        });
    }

    storeReducer(storePathName, reducer) {
        this.reducers = this.addValueByDottedPath(this.reducers, storePathName, reducer);
    }

    addValueByDottedPath(object, path, value = null) {
        const returnValue = { ...object };
        let curr = returnValue;
        this.forEachPathSegment(path, (segment, isLast) => {
            if (isLast) {
                if (value) {
                    curr[segment] = value;
                }
                return;
            }
            if (!curr[segment]) {
                curr[segment] = {};
            }
            curr = curr[segment];
        });
        return returnValue;
    }

    forEachPathSegment(path, callback) {
        const pathSegments = path.split('.');
        const currSegments = [];
        pathSegments.forEach((segment, index) => {
            currSegments.push(segment);
            return callback(segment, index === pathSegments.length - 1, currSegments.join('.'));
        });
    }

    retrieveReducer(storePathName) {
        return storePathName.split('.').reduce((o, i) => o[i], this.reducers);
    }
}
