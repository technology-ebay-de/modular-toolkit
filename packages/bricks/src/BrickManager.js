import { addValueByDottedPath, walkObject, getValueByDottedPath, filterObject, mergeObjects } from './utils';
import { registerSelectorsForUseWithGlobalState } from '@modular-toolkit/selectors';

/* exported for testing only */
export const addReducer = Symbol(
    process.env.NODE_ENV === 'production' ? undefined : 'addReducer (private method of BrickManager)'
);

const addSelectors = Symbol(
    process.env.NODE_ENV === 'production' ? undefined : 'addSelectors (private method of BrickManager)'
);

const addSaga = Symbol(process.env.NODE_ENV === 'production' ? undefined : 'addSaga (private method of BrickManager)');

const createInitialReducer = Symbol(
    process.env.NODE_ENV === 'production' ? undefined : 'createInitialReducer (private method of BrickManager)'
);

const saveReducer = Symbol(
    process.env.NODE_ENV === 'production' ? undefined : 'saveReducer (private method of BrickManager)'
);

const loadReducer = Symbol(
    process.env.NODE_ENV === 'production' ? undefined : 'loadReducer (private method of BrickManager)'
);

const hasReducer = Symbol(
    process.env.NODE_ENV === 'production' ? undefined : 'hasReducer (private method of BrickManager)'
);

const PREP_STATE = 'modular-toolkit/PREP_STATE';

export default class {
    constructor({ store, reducer = s => s, sagaMiddleware }) {
        this.store = store;
        this.initialReducer = this[createInitialReducer](reducer);
        this.initialStateProps = Object.keys(store.getState());
        this.store.replaceReducer(this.initialReducer);
        this.reducers = {};
        this.sagaMiddleware = sagaMiddleware;
    }

    installBricks(bricks) {
        for (const [storePath, module] of Object.entries(bricks)) {
            this.installBrick(storePath, module);
        }
    }

    installBrick(storePath, { reducer, saga, selectors }) {
        if (this[hasReducer](storePath)) {
            return;
        }
        this[addReducer](storePath, reducer);
        this[addSelectors](storePath, selectors);
        this[addSaga](saga);
    }

    [addReducer](storePath, reducer) {
        this.store.dispatch({ type: PREP_STATE, storePath });
        this[saveReducer](storePath, reducer);

        this.store.replaceReducer((state = {}, action) => {
            const stateForInitialReducer = filterObject(state, this.initialStateProps);
            let stateAfterInitialReduction = this.initialReducer(stateForInitialReducer, action);
            let hasChanges = stateForInitialReducer !== stateAfterInitialReduction;

            // this is important â€“ if we don't merge the state with the state after
            // running through the initial reducer (i.e. the one the Redux store was
            // created with), it will kick out the state from the reducers added later
            // stateAfterInitialReduction = { ...state, ...stateAfterInitialReduction };
            stateAfterInitialReduction = mergeObjects(state, stateAfterInitialReduction);

            let finalState = { ...stateAfterInitialReduction };
            walkObject(stateAfterInitialReduction, (currentSubState, path) => {
                const currReducer = this[loadReducer](path);
                if (typeof currReducer !== 'function') {
                    return;
                }
                const nextSubState = currReducer(currentSubState, action);
                if (nextSubState !== currentSubState) {
                    finalState = addValueByDottedPath(finalState, path, nextSubState);
                    hasChanges = true;
                }
            });

            return hasChanges ? finalState : state;
        });
    }

    [addSelectors](storePath, selectors) {
        registerSelectorsForUseWithGlobalState(storePath, selectors);
    }

    [addSaga](saga) {
        this.sagaMiddleware.run(saga);
    }

    [createInitialReducer](reducer) {
        return (state = {}, action) => {
            switch (action.type) {
                case PREP_STATE:
                    return {
                        ...addValueByDottedPath(state, action.storePath, {}, false)
                    };
                default:
                    return reducer(state, action);
            }
        };
    }

    [saveReducer](storePath, reducer) {
        this.reducers = addValueByDottedPath(this.reducers, storePath, reducer);
    }

    [loadReducer](storePath) {
        return getValueByDottedPath(this.reducers, storePath);
    }

    [hasReducer](storePath) {
        return this[loadReducer](storePath) !== null;
    }
}
