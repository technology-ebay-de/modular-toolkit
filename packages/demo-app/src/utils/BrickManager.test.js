import BrickManager, { loadReducer, saveReducer } from './BrickManager';
import { createStore } from 'redux';

const reducer = (state = {}, { type }) => {
    switch (type) {
        case 'THUD':
            return {
                ...state,
                thud: 'THUD'
            };
        default:
            return state;
    }
};

let store;
const sagaMiddleware = {
    run() {}
};
function arghReducer(state = {}, { type }) {
    switch (type) {
        case 'ARGH':
            return { ...state, argh: 'ARGH' };
        default:
            return state;
    }
}

function faselReducer(state = {}, { type }) {
    switch (type) {
        case 'FASEL':
            return { ...state, basel: 'BASEL' };
        case 'COOKIE':
            return { ...state, basel: 'COOKIE' };
        default:
            return state;
    }
}

function xyzzyReducer(state = {}, { type, payload }) {
    switch (type) {
        case 'XYZZY':
            return { ...state, payload };
        default:
            return state;
    }
}

describe('When I create a brick manager', () => {
    let brickManager;
    beforeEach(() => {
        store = createStore(reducer);
        return (brickManager = new BrickManager({ store, reducer, sagaMiddleware }));
    });
    describe('and I add a reducer with a simple store path', () => {
        beforeEach(() => brickManager.addReducer('argh', arghReducer));
        describe('and I dispatch an action for the simple store path reducer', () => {
            beforeEach(() => store.dispatch({ type: 'ARGH' }));
            describe('the Redux state', () => it('is correct', () => expect(store.getState()).toMatchSnapshot()));
        });
    });
    describe('and I add a reducer with a complex store path', () => {
        beforeEach(() => brickManager.addReducer('bricks.fasel', faselReducer));
        describe('and I dispatch and action for the root reducer', () => {
            beforeEach(() => store.dispatch({ type: 'THUD' }));
            describe('the Redux state', () => {
                let state;
                beforeEach(() => (state = store.getState()));
                it('is correct', () => expect(state).toMatchSnapshot());
            });
        });
        describe('and I dispatch an action for the complex store path reducer', () => {
            beforeEach(() => store.dispatch({ type: 'FASEL' }));
            describe('the Redux state', () => {
                let state;
                beforeEach(() => (state = store.getState()));
                it('has a root property that corresponds to the first segment of the complex store path', () => {
                    expect(state).toHaveProperty('bricks');
                });
                it('is correct', () => expect(state).toMatchSnapshot());
            });
            describe('and I dispatch a different action', () => {
                beforeEach(() => store.dispatch({ type: 'COOKIE' }));
                describe('the Redux state', () => {
                    let state;
                    beforeEach(() => (state = store.getState()));
                    it('is correct', () => expect(state).toMatchSnapshot());
                });
            });
            describe('and I dispatch and action for the root reducer', () => {
                beforeEach(() => store.dispatch({ type: 'THUD' }));
                describe('the Redux state', () => {
                    let state;
                    beforeEach(() => (state = store.getState()));
                    it('is correct', () => expect(state).toMatchSnapshot());
                });
            });
        });
        describe('and I add another reducer with a complex store path', () => {
            beforeEach(() => brickManager.addReducer('bricks.xyzzy', xyzzyReducer));
            describe('and I dispatch an action for the first reducer I added', () => {
                beforeEach(() => store.dispatch({ type: 'FASEL' }));
                describe('the Redux state', () => {
                    let state;
                    beforeEach(() => (state = store.getState()));
                    it('is correct', () => expect(state).toMatchSnapshot());
                });
            });
            describe('and I dispatch an action for the second reducer I added', () => {
                beforeEach(() => store.dispatch({ type: 'XYZZY', payload: "hey ho let's go!" }));
                describe('the Redux state', () => {
                    let state;
                    beforeEach(() => (state = store.getState()));
                    it('is correct', () => expect(state).toMatchSnapshot());
                });
            });
        });
    });
    describe('and I save a reducer (internal function) with a simple path', () => {
        beforeEach(() => brickManager[saveReducer]('fred', 'FRED_REDUCER'));
        describe('the internal property reducers', () => {
            let reducers;
            beforeEach(() => (reducers = brickManager.reducers));
            it('has the correct value', () => expect(reducers).toMatchSnapshot());
        });
    });
    describe('and I save a reducer (internal function) with a complex path', () => {
        beforeEach(() => brickManager[saveReducer]('bar.qux.fred', 'FRED_REDUCER'));
        describe('the internal property reducers', () => {
            let reducers;
            beforeEach(() => (reducers = brickManager.reducers));
            it('has the correct value', () => expect(reducers).toMatchSnapshot());
        });
        describe('and I retrieve the reducer', () => {
            let result;
            beforeEach(() => (result = brickManager[loadReducer]('bar.qux.fred')));
            describe('the result', () => it('is correct', () => expect(result).toEqual('FRED_REDUCER')));
        });
        describe('and I try to retrieve a reducer for the second segment of the path', () => {
            let result;
            beforeEach(() => (result = brickManager[loadReducer]('bar.qux')));
            describe('the result', () => it('is correct', () => expect(result).toEqual({ fred: 'FRED_REDUCER' })));
        });
        describe('and I save another reducer at the first level of the reducers tree', () => {
            beforeEach(() => brickManager[saveReducer]('bar.thud', 'THUD_REDUCER'));
            describe('the internal property reducers', () => {
                let reducers;
                beforeEach(() => (reducers = brickManager.reducers));
                it('has the correct value', () => expect(reducers).toMatchSnapshot());
            });
            describe('and I retrieve the reducer', () => {
                let result;
                beforeEach(() => (result = brickManager[loadReducer]('bar.thud')));
                describe('the result', () => it('is correct', () => expect(result).toEqual('THUD_REDUCER')));
            });
            describe('and I save yet another reducer that overwrites the path of the first one', () => {
                beforeEach(() => brickManager[saveReducer]('bar.qux', 'QUX_REDUCER'));
                describe('the internal property reducers', () => {
                    let reducers;
                    beforeEach(() => (reducers = brickManager.reducers));
                    it('has the correct value', () => expect(reducers).toMatchSnapshot());
                });
                describe('and I retrieve the reducer', () => {
                    let result;
                    beforeEach(() => (result = brickManager[loadReducer]('bar.qux')));
                    describe('the result', () => it('is correct', () => expect(result).toEqual('QUX_REDUCER')));
                });
            });
        });
    });
});
