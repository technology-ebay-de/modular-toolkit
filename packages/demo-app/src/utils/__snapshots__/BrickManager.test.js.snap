// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I create a brick manager and I install a Brick the reducer is stored 1`] = `
Object {
  "bricks": Object {
    "fasel": [Function],
  },
}
`;

exports[`When I create a brick manager and I install a brick with a reducer with a complex store path and I dispatch an action for the complex store path reducer and I dispatch a different action the Redux state is correct 1`] = `
Object {
  "bricks": Object {
    "fasel": Object {
      "basel": "COOKIE",
    },
  },
}
`;

exports[`When I create a brick manager and I install a brick with a reducer with a complex store path and I dispatch an action for the complex store path reducer and I dispatch and action for the root reducer the Redux state is correct 1`] = `
Object {
  "bricks": Object {
    "fasel": Object {
      "basel": "BASEL",
    },
  },
  "thud": "THUD",
}
`;

exports[`When I create a brick manager and I install a brick with a reducer with a complex store path and I dispatch an action for the complex store path reducer the Redux state is correct 1`] = `
Object {
  "bricks": Object {
    "fasel": Object {
      "basel": "BASEL",
    },
  },
}
`;

exports[`When I create a brick manager and I install a brick with a reducer with a complex store path and I dispatch and action for the root reducer the Redux state is correct 1`] = `
Object {
  "bricks": Object {
    "fasel": Object {},
  },
  "thud": "THUD",
}
`;

exports[`When I create a brick manager and I install a brick with a reducer with a complex store path and I install another brick with a reducer with a complex store path and I dispatch an action for the first reducer I added the Redux state is correct 1`] = `
Object {
  "bricks": Object {
    "fasel": Object {
      "basel": "BASEL",
    },
    "xyzzy": Object {},
  },
}
`;

exports[`When I create a brick manager and I install a brick with a reducer with a complex store path and I install another brick with a reducer with a complex store path and I dispatch an action for the second reducer I added the Redux state is correct 1`] = `
Object {
  "bricks": Object {
    "fasel": Object {},
    "xyzzy": Object {
      "payload": "hey ho let's go!",
    },
  },
}
`;

exports[`When I create a brick manager and I install a brick with a reducer with a simple store path and I dispatch an action for the simple store path reducer the Redux state is correct 1`] = `
Object {
  "argh": Object {
    "argh": "ARGH",
  },
}
`;

exports[`When I create a brick manager and I install two Bricks the reducers of both bricks are stored 1`] = `
Object {
  "bricks": Object {
    "fasel": [Function],
    "xyzzy": [Function],
  },
}
`;
